### DISCUSSION TITLE
Question about the extractor in the `tokio-postgres` example

### DISCUSSION THREAD

**Discussion by Moicen** [tokio-rs/axum#1757] [Q&A] [answered]
While I’m reading the example:

- [axum/main.rs at main · tokio-rs/axum · GitHub](https://github.com/tokio-rs/axum/blob/main/examples/tokio-postgres/src/main.rs#L53)

I found this part of code magical:

```rust
type ConnectionPool = Pool<PostgresConnectionManager<NoTls>>;

...

#[async_trait]
impl<S> FromRequestParts<S> for DatabaseConnection
where
    ConnectionPool: FromRef<S>,
    S: Send + Sync,
{
    type Rejection = (StatusCode, String);

    async fn from_request_parts(_parts: &mut Parts, state: &S) -> Result<Self, Self::Rejection> {
        let pool = ConnectionPool::from_ref(state);

        let conn = pool.get_owned().await.map_err(internal_error)?;

        Ok(Self(conn))
    }
}
```

I’m not sure how the above `ConnectionPool` type works. It seems in the implementation, the type is redefined as `ConnectionPool: FromRef<S>`, but it works. Is this a Rust language feature(I didn’t see similar usages in other places like this)?

**Comment by davidpdrsn** [comment, marked as answer]
That's just a regular trait bound.

**Reply by Moicen** [reply]
Ok now I see, because my type definition:

```rust
type MyDbState = Arc<DbState>;
```

Has implemented `Send+Sync` trait bound. In addition, I checked the code of `FromRef`:

```rust
pub trait FromRef<T> {
    /// Converts to this type from a reference to the input type.
    fn from_ref(input: &T) -> Self;
}

impl<T> FromRef<T> for T
where
    T: Clone,
{
    fn from_ref(input: &T) -> Self {
        input.clone()
    }
}
```

Because it implements itself for `T`, so all the trait bounds are met.

Thanks for help!
